# User Manual
This program is designed to take in 4 sequence alignments of genes within a bacteria type in fasta format and MIC antibiotic resistance values in a csv file.  Download the python code and sequence alignments for the test data.  The names within the code should be altered to their location on your machine.  To run this code, open it in an IDE and run it or open a command prompt with python installed and use "python mic_alleles.py" to run the code.

The names of the files within the code may be changed according to where they are located on your machine.  In order to run this code for  staphylococcus aureus, switch out the names of staphylococcus haemolyticus files.  The outfile names are also changable to what ever is desired.

This code is designed to take in four sequence alignments for 4 genes and MIC values in a csv format.

Mutations in the sequence alignment are found and compared to MIC values to determine which mutation(s) have a greater effect on MIC values.
